### DISCUSSION TITLE
Beta.19 New Features - Gotta go fast! 🏎💨

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#312] [Announcements]
Beta.19 is all about _speed_! I've finally landed on an approach I love to get the prompt moving waaaay faster.

Couple videos below:

## Instant Prompts

https://user-images.githubusercontent.com/36073/121084054-6d270a00-c79d-11eb-8b37-96473de7e0e4.mp4

```js
// Shortcut: option 5

let { items } = await db(async () => {
  let response = await get(
    `https://api.github.com/users/johnlindquist/repos`
  )

  return response.data
})

await arg("Select repo", items)

```

## Instant Tabs

https://user-images.githubusercontent.com/36073/121084134-85972480-c79d-11eb-9e18-94e5a2efa5d1.mp4

## Instant Main Menu

The main menu now also leverages the concepts behind Instant Prompts listed above.

## Faster in the future

These conventions laid the groundwork for caching prompt data, but I still have plenty ideas to speed things, especially around how the app launches the process. I'm looking forward to making this even faster for you!

I'm also starting the work on an "Instant Textarea" because I know popping open a little textarea to take/save notes/ideas is something many people would use. 📝

**Comment by cfritze64** [comment]
Nifty! Re the instant prompt: in an example like you show where the items are pulled from an api - each run of the script makes a new call to the api, right? So scriptkit shows what’s in the _scriptname.json file even while the script (moments later) calls out to the api, correct? And does scriptkit reconcile/update the file with new additions/deletions that the api call reveals?

**Reply by johnlindquist** [reply]
The `db` helper only runs the function if db .json file doesn't exist. So future runs won't call the API. The user becomes responsible to update the db using the `items.push()` and `await write()`.

Compare this (instant prompt with `db` helper, only makes request the first time the script runs):

```js
// Shortcut: option 5

let getRepos = async () => {
  let response = await get(
    `https://api.github.com/users/johnlindquist/repos`
  )

  return response.data
}

let { items } = await db(getRepos)

await arg("Select repo", items)
```

To this (using the same fn, but inside of an arg, so API request is made on every run). The items can't be displayed instantly:

```js
// Shortcut: option 5

let getRepos = async () => {
  let response = await get(
    `https://api.github.com/users/johnlindquist/repos`
  )

  return response.data
}

await arg("Select repo", getRepos)
```

It's essentially a caching strategy where the Kit.app is aware of the `_script-name.json` convention.

**Reply by zacjones93** [reply]
Wow I love this. I was going to look into implementing this functionality with a script I've been wanting to write so this came just in time 🙌

**Comment by kentcdodds** [comment]
It's is very cool! Yesterday I was looking into adding a db for my projects chooser thing but then when I timed it I found that loading them on demand only took 30ms 😅 So not really a great use case for a cache in my situation. But this is definitely simpler when a db is needed 👍 Thanks!
